accord-core/src/main/java/accord/impl/InMemoryCommandStore.java

```diff
@@ -92,6 +94,11 @@ public abstract class InMemoryCommandStore extends CommandStore
 
     private final TreeMap<TxnId, RangeCommand> rangeCommands = new TreeMap<>();
     private final TreeMap<TxnId, Ranges> historicalRangeCommands = new TreeMap<>();
+    /**
+     * Since this cache is fully in-memory the store does not hit states that most stores will; that data is not in-memory!
+     * To simulate such behaviors, a "cache" is used to state what is in-memory vs what needs to be loaded.
+     */
+    private final Set<Object> cache = new LinkedHashSet<>();
```

Comment:
this was unrelated to add/remove keyspace patch, it was mostly due to an observation we had a lot of code checking if things are cached to avoid loading and since this class is 100% in-memory we were at 100% cache hit rate, so went with faking a cache miss to get better coverage.

Reply:
We have a lot of places that try to access data outside of the `PreLoadContext` and acts on the global cache.

`accord.impl.InMemoryCommandStore#ifPresent(accord.primitives.TxnId)`
`accord.impl.InMemoryCommandStore.InMemorySafeStore#getIfLoaded(accord.primitives.TxnId)`
`accord.impl.AbstractSafeCommandStore#getInternalIfLoadedAndInitialised`
`accord.local.SafeCommandStore#ifLoadedAndInitialised`
`accord.local.Commands.NotifyWaitingOn#accept`

```
SafeCommand curSafe = ifLoadedAndInitialised(safeStore, depth);
                Command cur = curSafe != null ? curSafe.current() : null;
                if (cur == null)
                {
                    // need to load; schedule execution for later
                    safeStore.commandStore().execute(this, this).begin(safeStore.agent());
                    return;
                }
```

Cases like this would be 100% cache hit in BurnTest, so the non-cache hit case was never tested.

Reply:
s/a lot/a few/g

Reply:
I'll try to understand `commandStore.canExposeUnloaded()` and get back to you.  This looks to be a store property and not based off the requested data...

I added the cache logic as I noticed that the Cassandra logic was drastically different than InMemory logic, and we kept having logic optimized for InMemory logic... so I wanted to get the 2 stores closer as majority of testing is in BurnTest and far less is done with Cassandra...

Reply:
ill need to look closer at this change, after rebase we are failing and looks like its a cache hit issue... looking at the patch now (a long time remove from when I wrote it) I see we do the following

```
private static <K, V> V getIfLoaded(K key, Function<K, V> get, Consumer<V> add, Function<K, V> getIfLoaded)
    {
        V value = get.apply(key);
        if (value != null)
            return value;

        value = getIfLoaded.apply(key);
        if (value == null)
            return null;
        add.accept(value);
        return value;
    }
```

where `get` is defined as

```
protected InMemorySafeCommand getCommandInternal(TxnId txnId)
        {
            return commands.get(txnId);
        }
```

and `getIfLoaded` is

```
protected InMemorySafeCommand getIfLoaded(TxnId txnId)
        {
            GlobalCommand global = commandStore.ifPresent(txnId);
            return global != null ? global.createSafeReference() : null;
        }
```

so the `get` logic just reaches into the cache... so the `getIfLoaded` only happens iff `get` returned `null`... but that is just

```
public GlobalCommand ifPresent(TxnId txnId)
    {
        return cacheHit(txnId) ? commands.get(txnId) : null;
    }
```

so not sure how that matters at all...

Reply:
honestly I am fine pushing this into its own stand alone patch, not related to the core changes, so if can get the rest in faster, thats bette